### DISCUSSION TITLE
Should spatial vectors be required to have geometries?

### DISCUSSION THREAD

**Discussion by phargogh** [natcap/invest#1305] [General]
InVEST has, for quite some time, assumed that if we have a spatial vector layer[1], all features in the layer must have geometries.  In some, but not all, cases, these geometries must also be valid.

Before just jumping into attempting to solve this, it would be helpful to see how we handle these cases throughout InVEST and pygeoprocessing.  Are there examples where geometries are not required to be present?  And is this something that is reasonable to specify in `ARGS_SPEC` and validate in `validation`?

This issue came up while [debugging an issue on the forums](https://community.naturalcapitalproject.org/t/coastal-vulnerability-3-8-0-model-attribute-error-troubleshooting/885/2), but it turned into a broader question warranting an issue.

[1] Note that, especially with GeoPackages, it's [possible to create an aspatial layer](https://gist.github.com/phargogh/f654a8de747544272f356fb34024dc87#file-aspatial-py).

**Comment by phargogh** [comment]
This might be similar to the undefined nodata issue: if it's supported, we should handle it.

We might already have a generator in a couple places in InVEST that does this (skip missing geometries and also fix invalid geometries).  Used in at least recreation and CV, and in a few places in each one.

**Comment by davemfish** [comment]
> This might be similar to the undefined nodata issue: if it's supported, we should handle it.

I tend to agree with this. Skipping & warning about missing geometries is a better user-experience than crashing.
